Setting system mode for Eurotronic thermostat

sti0 opened the pull request:
Setting the system mode for Eurotronic thermostat based on the current temperature set point which is published on temperature input changes.

5°C        = off
6°C - 29°C = heat
30°C       = boost / auto

There is also some work to do in toZigbee.js to set the right mode.

Prose code:
  - if system mode = off then publish 32 to 0x4008
  - if system mode = auto then publish 5 to 0x4008
  - if system mode = heat then publish 17 (?) to 0x4008

=> key: 'system_mode' must be set on another way as it is on the generic implementation. Because there is already a eurotronic_system_mode which allows more than setting a state (also set child protection and mirror the display). I suggest to name it "eurotronic_thermostat_system_mode".

I've done some tests but I'm not getting any further. @Koenkk could you point me in a direction?

sti0 commented:
![image](https://user-images.githubusercontent.com/18382402/66708812-fc74e380-ed56-11e9-89a6-ad784ff28fe5.png)
![image](https://user-images.githubusercontent.com/18382402/66708814-03035b00-ed57-11e9-87cc-13530eb0a39c.png)
![image](https://user-images.githubusercontent.com/18382402/66708815-07c80f00-ed57-11e9-848c-a6c0627a77c1.png)

allofmex commented:
What is your goal here? Are you trying to workaround the issue that eurotronics seems not to make use of standard `system_mode`? (my thermostats seems to report always 4 (heating) in attribute systemMode(id 28), no matter if in boost or not)

In that case (and if it is really the goal to overwrite device's report by our own interpretation!), you should set system_mode based on reported bits in `eurotronic_system_mode`! I don't think it is a good idea to base it on the temperature value since normally you can set 5 or 30 degree even if device is not in off or boost mode!

sti0 commented:
First of all this is not my interpretation, it's the way the device acts. If I put it in boost mode (`eurotronic_system_mode `= 5), it responses with a `current_temperature_setpoint` = 30. If I turn it off (`eurotronic_system_mode `= 32) it responses with a `current_temperature_setpoint `= 5.

Now to my goals. I would like to get a `system_mode `for this device. While I'm using Home Assistant, the state (=`system_mode`) is actually not published to HA. You are absolutely right, setting `system_mode `based on the reported `eurotronic_system_mode `would be the perfect deal. But in my tests I set the `eurotronic_system_mode `but the device responses with an old or different code. 

Here's an example:
Initial message, normal heating mode (`eurotronic_system_mode `= 1):
`zigbee2mqtt:info  2019-10-13T18:12:41: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":16.5,"eurotronic_system_mode":1,"local_temperature":20.25,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":136,"system_mode":"heat"}'`

Now turing it off by setting `eurotronic_system_mode `= 32:
`zigbee2mqtt:info  2019-10-13T18:14:25: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":5,"eurotronic_system_mode":17,"local_temperature":20.25,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":136,"system_mode":"off"}'`
=> Device shows "off" at the display. `current_heating_setpoint `= 5, `eurotronic_system_mode` = 17

If I now reactivate it by setting eurotronic_system_mode = 17, to retrieve the former state:
`zigbee2mqtt:info  2019-10-13T18:16:32: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":16.5,"eurotronic_system_mode":1,"local_temperature":20.25,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":136,"system_mode":"heat"}'`
=> device respones with `eurotronic_system_mode `= 1 and setting `current_heating_setpoint `back to 16.5

If I now put it into boost mode, `eurotronic_system_mode `= 5
`zigbee2mqtt:info  2019-10-13T18:18:42: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":30,"eurotronic_system_mode":5,"local_temperature":20.25,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":255,"battery":100,"linkquality":136,"system_mode":"auto"}'`
=> `system_mode `was set to 5, `current_heating_setpoint `changes to 30 degrees

But if I turn it off now (eurotronic_system_mode = 32), display shows "off", but eurotronic_system_mode belongs to 5, while current_heating_setpoint changes to 5 degrees
`zigbee2mqtt:info  2019-10-13T18:19:49: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":5,"eurotronic_system_mode":5,"local_temperature":20.25,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":255,"battery":100,"linkquality":136,"system_mode":"off"}'`

If I now put it from off to boost mode, `eurotronic_system_mode `changes to 17
zigbee2mqtt:info  2019-10-13T18:21:03: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":30,"eurotronic_system_mode":17,"local_temperature":20.25,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":136,"system_mode":"auto"}'
=> `current_heating_setpoint `changes to 30

Maybe there is a better way but setting system_mode along to eurotronic_system_mode is not really reliable. Therefore I handled it this way and setting system_mode along to current_heating_setpoint which was reliable in my tests. The only downside is if you change the degree to 5 or 30 it pushes a system_mode (off or auto) and the display shows the degrees instead of on or off. This is a small deviation which I can live with.

sti0 commented:
Here's a second log including the data output received by herdsman (row before zigbee2mqtt msg). The test was done from ground up (shut down the device by removing the batteries shortly).

Please note that 16387 = 0x4003 belongs to the current_heating_setpoint (bold) and 16392 = 0x4008 belongs to the eurotronic_system_mode (italic).

### Let's get started:

zigbee2mqtt:info  2019-10-13T19:28:00: Starting interview of 'climate_buro'
zigbee2mqtt:info  2019-10-13T19:28:00: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"pairing","message":"interview_started","meta":{"friendly_name":"climate_buro"}}'
{ '16387': 2000 }
zigbee2mqtt:info  2019-10-13T19:28:00: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":20,"eurotronic_system_mode":1,"local_temperature":22.11,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":128,"system_mode":"heat"}'
{ localTemp: 2000 }
zigbee2mqtt:info  2019-10-13T19:28:00: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":20,"eurotronic_system_mode":1,"local_temperature":20,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":131,"system_mode":"heat"}'
zigbee2mqtt:info  2019-10-13T19:28:01: Successfully interviewed 'climate_buro', device has successfully been paired
zigbee2mqtt:info  2019-10-13T19:28:01: Device 'climate_buro' is supported, identified as: Eurotronic Spirit Zigbee wireless heater thermostat (SPZB0001)
zigbee2mqtt:info  2019-10-13T19:28:01: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"pairing","message":"interview_successful","meta":{"friendly_name":"climate_buro","model":"SPZB0001","vendor":"Eurotronic","description":"Spirit Zigbee wireless heater thermostat","supported":true}}'
{ '16387': 1600 }
zigbee2mqtt:info  2019-10-13T19:28:02: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":16,"eurotronic_system_mode":1,"local_temperature":20,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":123,"system_mode":"heat"}'
{ localTemp: 2211 }
zigbee2mqtt:info  2019-10-13T19:28:02: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":16,"eurotronic_system_mode":1,"local_temperature":22.11,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":123,"system_mode":"heat"}'

### => turing into boost mode {"eurotronic_system_mode":5}
**{ '16387': 3000 }
zigbee2mqtt:info  2019-10-13T19:29:06: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":30,"eurotronic_system_mode":1,"local_temperature":22.11,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":136,"system_mode":"auto"}'**
_{ '16392': 5 }
zigbee2mqtt:info  2019-10-13T19:30:06: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":30,"eurotronic_system_mode":5,"local_temperature":22.11,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":134,"system_mode":"auto"}'_
{ unoccupiedHeatingSetpoint: 700,
  occupiedHeatingSetpoint: 1500,
  pIHeatingDemand: 252 }
zigbee2mqtt:info  2019-10-13T19:30:06: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":30,"eurotronic_system_mode":5,"local_temperature":22.11,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":252,"battery":100,"linkquality":134,"system_mode":"auto"}'

### => back to normal heating {"eurotronic_system_mode":17}
**{ '16387': 1600 }
zigbee2mqtt:info  2019-10-13T19:30:43: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":16,"eurotronic_system_mode":5,"local_temperature":22.11,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":252,"battery":100,"linkquality":134,"system_mode":"heat"}'**
_{ '16392': 1 }
zigbee2mqtt:info  2019-10-13T19:32:07: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":16,"eurotronic_system_mode":1,"local_temperature":22.11,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":252,"battery":100,"linkquality":136,"system_mode":"heat"}'_
{ pIHeatingDemand: 0 }
zigbee2mqtt:info  2019-10-13T19:32:07: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":16,"eurotronic_system_mode":1,"local_temperature":22.11,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":136,"system_mode":"heat"}'

**=> now turn it off {"eurotronic_system_mode":32}**
**{ '16387': 500 }
zigbee2mqtt:info  2019-10-13T19:32:26: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":5,"eurotronic_system_mode":1,"local_temperature":22.11,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":134,"system_mode":"off"}'**
_{ '16392': 17 }
zigbee2mqtt:info  2019-10-13T19:34:08: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":5,"eurotronic_system_mode":17,"local_temperature":22.11,"occupied_heating_setpoint":15,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":134,"system_mode":"off"}'_

As you can see the `current_heating_setpoint` is send first every time and in near real-time. The `eurotronic_system_mode `change is published after a minimum of 1 minute. In this experiment I waited for the `eurotronic_system_mode `message to be set finally before I changed the state of the device. If you change it faster than it will end up in a completely wrong state which you can see in my prior comment.

sti0 commented:
Added the "way back" now. If system_mode is set to off, heat or auto it will set right state to attribute 0x4008 (eurotronic_system_mode).

On my point of view this change is ready for a merge.

allofmex commented:
Thanks for this detailed tests.

Found the problem:
There is currently no reporting configured for eurotronic_system_mode, so it updates only every 10 minutes or so.

Please add in devices.js -> Eurotronic -> configure
```
(cb) => device.foundation('hvacThermostat', 'configReport', [{ //SPBZ0001_system_mode
             direction: 0, attrId: 0x4008, dataType: 34, minRepIntval: 0,
             maxRepIntval: 600, repChange: 1}], {manufSpec: 1, manufCode: 4151}, cb),
```

Then `eurotronic_system_mode` updates immediately after any change. I tested it and shows the correct values now.
Now you can update normal system_mode based on bits set in eurotronic_system_mode instead of temperature values.

Please also take care that you use the single bits only and not the whole byte! One reason: Bit 7 is child-protection, that may be set extra to other bits. (Example: boost without child-protection is 00000101=5, with c-p is 10000101=133)

Someone said that Bit 5 (32) is Window-Open! He mentioned that it gets reset to normal mode after ? minutes! I waited 10 minutes but nothing happened. Maybe it waits for a raising temperature. Not sure about this.

sti0 commented:
Ok thanks for this information. Will implement this and come back to you.

Just for information I set eurotronic_system_mode to 32 and it stays off overnight. Don't think this would be set back to heating mode automatically.

sti0 commented:
I'm implemented your advice in current zigbee-herdsman-converter branch (zigbee2mqtt edge) like this:
```
// Eurotronic
    {
        zigbeeModel: ['SPZB0001'],
        model: 'SPZB0001',
        vendor: 'Eurotronic',
        description: 'Spirit Zigbee wireless heater thermostat',
        supports: 'temperature, heating system control',
        fromZigbee: [
            fz.thermostat_att_report,
            fz.eurotronic_thermostat,
            fz.generic_battery_remaining,
        ],
        toZigbee: [
            tz.thermostat_occupied_heating_setpoint, tz.thermostat_unoccupied_heating_setpoint,
            tz.thermostat_local_temperature_calibration, tz.thermostat_system_mode,
            tz.eurotronic_system_mode, tz.eurotronic_error_status, tz.thermostat_setpoint_raise_lower,
            tz.thermostat_control_sequence_of_operation, tz.thermostat_remote_sensing,
            tz.eurotronic_current_heating_setpoint, tz.eurotronic_trv_mode, tz.eurotronic_valve_position,
        ],
        meta: {configureKey: 1},
        configure: async (device, coordinatorEndpoint) => {
            const endpoint = device.getEndpoint(1);
            await bind(endpoint, coordinatorEndpoint, ['genPowerCfg', 'hvacThermostat']);
            await configureReporting.thermostatTemperature(endpoint);
            await endpoint.configureReporting('hvacThermostat', [{
                attribute: {ID: 0x4003, type: 41},
                minimumReportInterval: 0,
                maximumReportInterval: repInterval.HOUR,
                reportableChange: 25,
            }]);
            await endpoint.configureReporting('hvacThermostat', [{
                attribute: {ID: 0x4008, type: 34},
                minimumReportInterval: 0,
                maximumReportInterval: repInterval.HOUR,
                reportableChange: 1,
            }]);
        },
    },
```
Slightly other syntax. But it's not reporting `eurotronic_system_mode `at all. So I changed the type to 36 (0x24 = 36, not 34 maybe a typo on your code).

With this type 36 it changes the eurotronic_system_mode but not immediatly at all:

Short example:
Turning from normal heat to off:
```
{ '16387': 500 }
zigbee2mqtt:info  2019-10-14T20:01:08: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":5,"eurotronic_system_mode":1,"local_temperature":21.22,"occupied_heating_setpoint":12,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":128,"system_mode":"off"}'
{ '16392': 17 }
zigbee2mqtt:info  2019-10-14T20:02:08: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":5,"eurotronic_system_mode":17,"local_temperature":21.22,"occupied_heating_setpoint":12,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":128,"system_mode":"off"}'
```

=> `eurotronic_system_mode `changes 1 minute after the `current_heating_setpoint `was exposed.

In fact, this doesn't change anything. Maybe my implementation is wrong? What does the `reportableChange `attribute (for `current_heating_setpoint `it's set to 25)?

sti0 commented:
And another one:
Restarted zigbee2mqtt, both values get delivered on the same time (changing off to heat)
```
{ '16387': 1400, '16392': 1 }
zigbee2mqtt:info  2019-10-14T20:07:57: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":14,"eurotronic_system_mode":1,"local_temperature":21.22,"occupied_heating_setpoint":12,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":128,"system_mode":"off"}'
```

but changing back to off, eurotronic_system_mode arrives nearly 2 minutes later
```
{ '16387': 500 }
zigbee2mqtt:info  2019-10-14T20:08:10: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":5,"eurotronic_system_mode":1,"local_temperature":21.22,"occupied_heating_setpoint":12,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":131,"system_mode":"off"}'
{ '16392': 17 }
zigbee2mqtt:info  2019-10-14T20:10:04: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":5,"eurotronic_system_mode":17,"local_temperature":21.22,"occupied_heating_setpoint":12,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":131,"system_mode":"off"}'
```

@allofmex Could you implement your solution on the herdsman converters and test it?

sti0 commented:
Even if I remove the existing implementation of reporting current_hreating_setpoint
```
await endpoint.configureReporting('hvacThermostat', [{
                attribute: {ID: 0x4003, type: 41},
                minimumReportInterval: 0,
                maximumReportInterval: repInterval.HOUR,
                reportableChange: 25,
            }]);
```

The result is the same (off -> heat -> off):
```
{ '16387': 1400, '16392': 1 }
zigbee2mqtt:info  2019-10-14T20:25:25: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":14,"eurotronic_system_mode":1,"local_temperature":21.58,"occupied_heating_setpoint":12,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":128,"system_mode":"off"}'
{ '16387': 500 }
zigbee2mqtt:info  2019-10-14T20:25:38: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":5,"eurotronic_system_mode":1,"local_temperature":21.58,"occupied_heating_setpoint":12,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":131,"system_mode":"off"}'
{ '16392': 17 }
zigbee2mqtt:info  2019-10-14T20:27:32: MQTT publish: topic 'zigbee2mqtt/climate_buro', payload '{"current_heating_setpoint":5,"eurotronic_system_mode":17,"local_temperature":21.58,"occupied_heating_setpoint":12,"unoccupied_heating_setpoint":7,"eurotronic_error_status":0,"pi_heating_demand":0,"battery":100,"linkquality":131,"system_mode":"off"}'
```

allofmex commented:
I am not using herdsman yet (iobroker is still on shepherd-converter).
Looks for me like `{manufSpec: 1, manufCode: 4151}` is missing (for 0x4003 and 0x4008). This is important, since both attributes are manufacturer specific attributes and usually the response is unsupAttribute if not addressed with manufSpec...

sti0 commented:
@Koenkk how to set manufacture specific attributes in herdsman?

Koenkk commented:
```js
await endpoint.configureReporting(
    'hvacThermostat', 
    [{attribute: {ID: 0x4003, type: 41}, minimumReportInterval: 0, maximumReportInterval: repInterval.HOUR, reportableChange: 25}],
    {manufacturerCode: 4151}
);
```

sti0 commented:
Thanks @Koenkk ! I added the manufacturerCode but still same behavior...temperature gets published immediately, system mode after some minutes...

```
// Eurotronic
    {
        zigbeeModel: ['SPZB0001'],
        model: 'SPZB0001',
        vendor: 'Eurotronic',
        description: 'Spirit Zigbee wireless heater thermostat',
        supports: 'temperature, heating system control',
        fromZigbee: [
            fz.thermostat_att_report,
            fz.eurotronic_thermostat,
            fz.generic_battery_remaining,
        ],
        toZigbee: [
            tz.thermostat_occupied_heating_setpoint, tz.thermostat_unoccupied_heating_setpoint,
            tz.thermostat_local_temperature_calibration, tz.thermostat_system_mode,
            tz.eurotronic_system_mode, tz.eurotronic_error_status, tz.thermostat_setpoint_raise_lower,
            tz.thermostat_control_sequence_of_operation, tz.thermostat_remote_sensing,
            tz.eurotronic_current_heating_setpoint, tz.eurotronic_trv_mode, tz.eurotronic_valve_position,
        ],
        meta: {configureKey: 1},
        configure: async (device, coordinatorEndpoint) => {
            const endpoint = device.getEndpoint(1);
            await bind(endpoint, coordinatorEndpoint, ['genPowerCfg', 'hvacThermostat']);
            await configureReporting.thermostatTemperature(endpoint);
            const options = {manufacturerCode: 4151};
            await endpoint.configureReporting('hvacThermostat', [{
                attribute: {ID: 0x4003, type: 41},
                minimumReportInterval: 0,
                maximumReportInterval: repInterval.HOUR,
                reportableChange: 25,
            }], options);
            await endpoint.configureReporting('hvacThermostat', [{
                attribute: {ID: 0x4008, type: 36},
                minimumReportInterval: 0,
                maximumReportInterval: repInterval.HOUR,
                reportableChange: 1,
            }], options);
        },
    },
```

Koenkk commented:
Can you try reduce the `reportableChange` to `1`?

sti0 commented:
Changed the `reportableChange `to 1 for 0x4003 (=> current_heating_setpoint), but still same behavior. :-(

allofmex commented:
> {**manufSpec: 1**, manufCode: 4151}

manufSpec is missing!

sti0 commented:
On herdsman there is no attribute manufSpec anymore....

https://github.com/Koenkk/zigbee-herdsman/blob/master/docs/api/interfaces/_controller_model_endpoint_.options.md

Koenkk commented:
Manufspec is not needed, it will automatically be switched if a manufacturerCode is provided.

sti0 commented:
Could this maybe have something to do with the reportable 'N' flag in the Eurotronic docs? Current Heating Setpoint's reportable flag is 'Y' in the docs.
![image](https://user-images.githubusercontent.com/18382402/66868054-e5bfce00-ef9c-11e9-923a-7132b8dbbbc8.png)

https://eurotronic.org/wp-content/uploads/2019/01/Spirit_ZigBee_BAL_web_DE_view_V9.pdf

allofmex commented:
Tested again with my device:
I **removed** configure for 0x4008, factory reset device (all buttons pressed), restarted broker
eurotronic_system_mode does **not** update at changes, as you describe.

Then I added configuration again
```
 (cb) => device.foundation('hvacThermostat', 'configReport', [{ //SPBZ0001_system_mode
                    direction: 0, attrId: 0x4008, dataType: 34, minRepIntval: 0,
                    maxRepIntval: 600, repChange: 1}], {manufSpec: 1, manufCode: 4151}, cb),
```
Restart broker, **power-reset (remove battery) of device to force instant rejoin/reconfigure**, waiting until configuration process appears in logs:

Result: instant reporting of eurotronic_system_mode **works** with zigbee-shepherd-converter.

- I press the boost button, seconds later eurotronic_system_mode changes to 5.
- Press button again, mode changes to 1.
- Send 32, mode changes to 17 (don't get Eurotronics logic here, why they use bit 4 and 5, but that's different topic)
- Send 17, mode changes to 1.


`2019-10-16 10:15:42.619 Zigbee publish to '0x00158d000192d12c', hvacThermostat - cmd write - payload [{"attrId":16392,"dataType":34,"attrData":32}] - cfg {"manufSpec":1,"manufCode":4151} - endpoint 1`
`2019-10-16 10:15:45.986 Device 0x00158d000192d12c emit event attReport with data:{"cid":"hvacThermostat","data":{"16387":500,"16392":17}}`
`2019-10-16 10:15:45.986 Device 0x00158d000192d12c emit event devChange with data:{"cid":"hvacThermostat","data":{"16387":500,"16392":17}}`


All eurotronic_system_mode updates happen after latest a few seconds, so reporting config on 0x4008 definitely works.


Seems to be some issue with reporting configuration (in herdsman?).
What do your logs say about the configuration requests and there result?

Koenkk commented:
@sti0 after you change the `configure` method, you need to bump the `configureKey`, otherwise zigbee2mqtt won't reconfigure it. Did you do this?

Koenkk commented:
Expect bumping the `configureKey` for this device everytime nothing, so e.g. `meta: {configureKey: 1},` -> `meta: {configureKey: 2},`

sti0 commented:
This did the trick. Reverted all changes and implemented `system_mode `setting by `eurotronic_system_mode`

@Koenkk , @allofmex  thanks for your support!

Koenkk commented:
Is this ready to be merged?

allofmex commented:
Great. Thanks for not giving up :)

One remains:
As mentioned before, you should really handle only the bits, do not overwrite the whole bitmap!

fromZigbee
```
if ((result.eurotronic_system_mode & 1 << 4) != 0) { // bit 4 is set (note: gets set by bit 5)
      result.system_mode = common.thermostatSystemModes[0]; // off
} else if ((result.eurotronic_system_mode & 1 << 2) != 0) { // bit 2 is set
    result.system_mode = common.thermostatSystemModes[5]; // boost => auto
} 
...
```

toZigbee
```
switch (systemMode) {
case 0:
    value |= 1<<5; // set bit 5 = off (note: to be read as bit 4!)
    break;
case 5: // boost/emergency heating
    value |= 1<<2; // set bit 2 = boost
    break;
 ....
}
```
Your current implementation breaks at least child-protection setting (bit 7)!

sti0 commented:
Implemented the bitwise setting. Had to switch off and boost @fromZigbee.js because if you switch from off directly to boost it will publish mode 21 which matches on 1<<2 and 1<<4.

Koenkk commented:
@allofmex @sti0 let me know when this is ready for merge.

sti0 commented:
@allofmex 
I don't think that `boost `is an `emergency heating`. Feel free to read the following article about emergency heating.
> Emergency Heat, also known as Auxiliary Heat, is the second stage of heat that your thermostat utilizes when the temperature is too cold for your heat pump to remove heat from the outside. The outdoor temperature that usually triggers the Emergency Heat setting to turn on is 35°F and below.

https://www.bobsredtrucks.com/blog/what-is-emergency-heat/

Boost is just another form of heating with the highest temperature setpoint and should be deactivated after some time. `auto` maybe not the right state but while there is no `boost` state it's the closest one in my opinion.

@Koenkk 
I think this one is ready to merge. I tested the mode setting again and works flawless.

sti0 commented:
@Koenkk `temperature_calibration` is not working for this device. What could be the reason?

I added `temperature_calibration` to my devices.yaml like I did for Xiaomi temperature and Hue motion sensors. Is this feature not implemented for climates?

```
'0x00158d000192de9e':
  friendly_name: climate_buro
  retain: false
  temperature_calibration: 2
```

Koenkk commented:
It's not implemented yet, this can be done inside the converters (e.g. https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/converters/fromZigbee.js#L301). Example: https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/converters/fromZigbee.js#L379

Note that if you implement this in `fromZigbee` you also need to implement it in `toZigbee`.

allofmex commented:
Did I misunderstood something?
For me it seems to work by existing
toZigbee.js:
tz.thermostat_local_temperature_calibration

sti0 commented:
Because you don't use zigbee2mqtt its not relevant for you. ioBroker uses the underlying herdsman implementation. In zigbee2mqtt you could set a `temperature_calibration` in the device settings. This needs to be implemented in the converters like @Koenkk  said.

sti0 commented:
Added temperature_calibration option for all thermostats. Is it needed to put this in the docs or would this be added automatically? ("Device specific configuration" https://www.zigbee2mqtt.io/devices/WSDCGQ11LM.html)

Koenkk commented:
Great, I will make sure it ends up in the docs. Thanks!

allofmex commented:
> Because you don't use zigbee2mqtt its not relevant for you. ioBroker uses the underlying herdsman implementation. In zigbee2mqtt you could set a temperature_calibration in the device settings.

@sti0 (Sorry again for my complaints :-/ )
The `calibrateAndPrecisionRoundOptions` is an **offline** calibration! It is fine for temperature sensors and similar that do not have own calibration.

But if you use it for the thermostat, **only the reported temperature** will be calibrated (after it was received from device). The temperature the thermostat uses for opening/closing valve is **still uncalibrated**!!
Only if the offset is **written to device** by `tz.thermostat_local_temperature_calibration`, both (reported and internally used) will be calibrated!

Koenkk commented:
@allofmex you are right, I missed this when merging the PR. @sti0  Could you also implement the conversion for the `toZigbee` side? (https://github.com/Koenkk/zigbee-herdsman-converters/pull/682#issuecomment-544116133)

sti0 commented:
You're right. Think we could remove the calibration change. If setting local_temperature_calibration payload to the device. It gives the calibrated value back. If you set local_temperature_calibration and temperature_calibration in z2m it increases temperature twice. 
@Koenkk what do you think? I have no idea how to fix this in toZigbee.js

Koenkk commented:
If the device can also do it himself, then remove it from the converters and use `temperature_calibration`.

sti0 commented:
Reverted in #698
